contents/testing/writing-testable-code.md

```diff
@@ -0,0 +1,146 @@
+# Writing Testable Code
+
+When the project is big enough and needs to be maintainable in the long run, it has to rely on automated tests to keep its quality. Especially, the idea of unit testing becomes very important for Object-Oriented programs. Compared to system testing where you test the program altogether, unit testing has its benefit for being fast (because it only instantiates a small piece of the program) and stable (because it usually mocks out the unstable dependency, e.g. network connection, database connection).
+
+This article compiles multiple posts from [Google Testing Blog](https://testing.googleblog.com) on how to write testable code. It explains four common flaws in untestable code, and rules to follow for each of the flaws.
+
+## Flaw #1: No seams for isolating the class under test
+
+> Seams is where we prevent the execution of normal code path and is how we achieve isolation of the class under test. - [Google Testing Blog](https://testing.googleblog.com/2008/12/static-methods-are-death-to-testability.html)
+
+To create seams in your code, you need to acquire your dependency via [dependency injection](https://en.wikipedia.org/wiki/Dependency_injection). You inject real object for production and [test doubles](https://www.martinfowler.com/bliki/TestDouble.html) for testing. For example, if a class depends on some database connection to retrieve some user data, you can inject a fake database connection to always return prepared data immediately without really performing any database operation.
+
+That is the basic idea for how seams can help in unit testing. Below are five rules you can follow to create seams in your code.
+
+### Rule #1: Don't do actual work in constructors
+
+#### Warning Signs
+
+* `new` keyword in a constructor or at field declaration
+* Code does complex object graph construction inside a constructor rather than using a factory or builder
+* Anything more than field assignment in constructors
+
+> Any work you do in a constructor needs to successfully navigate through in every test (not just the direct test, but also any related test which tries to instantiate your class indirectly as part of some larger object graph). - [Google Testing Blog](https://testing.googleblog.com/2008/08/by-miko-hevery-so-you-decided-to.html)
+
+In short, the constructor's job is to assign the dependencies to fields. And that's all.
```

Comment:
The rule is Don't do actual work in constructors.

checking invariants hold is actually not considered as "actual work".